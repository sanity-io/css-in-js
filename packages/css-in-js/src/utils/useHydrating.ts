import {useSyncExternalStore} from 'react'

export function useHydrating(): boolean {
  return useSyncExternalStore(noop, getSnapshot, getServerSnapshot)
}

function noop() {
  return () => {}
}

function getSnapshot() {
  return false
}

function getServerSnapshot() {
  return true
}
