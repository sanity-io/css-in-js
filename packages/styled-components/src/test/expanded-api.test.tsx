import React from 'react';
import { render, fireEvent, screen, act } from '@testing-library/react';
import { getRenderedCSS, resetStyled } from './utils';

// Disable isStaticRules optimisation since we're not
// testing for ComponentStyle specifics here
vi.mock('../utils/isStaticRules', () => ({ default: () => false }));

let styled: ReturnType<typeof resetStyled>;

describe('expanded api', () => {
  /**
   * Make sure the setup is the same for every test
   */
  beforeEach(() => {
    styled = resetStyled();
  });

  describe('displayName', () => {
    it('should be auto-generated if none passed', () => {
      const Comp = styled.div``;
      expect(Comp.displayName).toBe('styled.div');
    });

    it('should be attached if supplied', () => {
      const Comp = styled.div.withConfig({ displayName: 'Comp' })``;
      expect(Comp.displayName).toBe('Comp');
    });
  });

  describe('componentId', () => {
    it('should be generated as "sc" + hash', () => {
      const Comp = styled.div``;
      const Comp2 = styled.div``;
      expect(Comp.styledComponentId).toBe('sc-a');
      expect(render(<Comp />).container).toMatchSnapshot();
      expect(Comp2.styledComponentId).toBe('sc-b');
      expect(render(<Comp2 />)).toMatchSnapshot();
    });

    it('should be generated from displayName + hash', () => {
      const Comp = styled.div.withConfig({ displayName: 'Comp' })``;
      const Comp2 = styled.div.withConfig({ displayName: 'Comp2' })``;
      expect(Comp.styledComponentId).toBe('Comp-a');
      expect(render(<Comp />).container).toMatchSnapshot();
      expect(Comp2.styledComponentId).toBe('Comp2-b');
      expect(render(<Comp2 />)).toMatchSnapshot();
    });

    it('should be attached if passed in', () => {
      const Comp = styled.div.withConfig({ componentId: 'LOLOMG' })``;
      const Comp2 = styled.div.withConfig({ componentId: 'OMGLOL' })``;
      expect(Comp.styledComponentId).toBe('LOLOMG');
      expect(render(<Comp />).container).toMatchSnapshot();
      expect(Comp2.styledComponentId).toBe('OMGLOL');
      expect(render(<Comp2 />)).toMatchSnapshot();
    });

    it('should be combined with displayName if both passed in', () => {
      const Comp = styled.div.withConfig({
        displayName: 'Comp',
        componentId: 'LOLOMG',
      })``;
      const Comp2 = styled.div.withConfig({
        displayName: 'Comp2',
        componentId: 'OMGLOL',
      })``;
      expect(Comp.styledComponentId).toBe('Comp-LOLOMG');
      expect(render(<Comp />).container).toMatchSnapshot();
      expect(Comp2.styledComponentId).toBe('Comp2-OMGLOL');
      expect(render(<Comp2 />).container).toMatchSnapshot();
    });
  });

  describe('chaining', () => {
    it('should merge the options strings', () => {
      const Comp = styled.div
        .withConfig({ componentId: 'id-1' })
        .withConfig({ displayName: 'dn-2' })``;
      expect(Comp.displayName).toBe('dn-2');
      expect(render(<Comp />).container).toMatchSnapshot();
    });
  });

  describe('"as" prop', () => {
    it('changes the rendered element type', () => {
      const Comp = styled.div`
        color: red;
      `;

      expect(render(<Comp as="span" />).container).toMatchSnapshot();
    });

    it('changes the rendered element type when used with attrs', () => {
      const Comp = styled.div.attrs(() => ({
        as: 'header',
      }))`
        color: red;
      `;

      expect(render(<Comp />).container).toMatchSnapshot();
    });

    it('prefers attrs over props', () => {
      const Comp = styled.div.attrs(() => ({
        as: 'header',
      }))`
        color: red;
      `;

      expect(render(<Comp as="span" />).container).toMatchInlineSnapshot(`
        <div>
          <header
            class="sc-a b"
          />
        </div>
      `);
    });

    it('works with custom components', () => {
      const Override: React.FC<any> = props => <figure {...props} />;
      const Comp = styled.div`
        color: red;
      `;

      expect(render(<Comp as={Override} />).container).toMatchSnapshot();
    });

    it('transfers all styles that have been applied', () => {
      const Comp = styled.div`
        background: blue;
        color: red;
      `;

      const Comp2 = styled(Comp)`
        color: green;
      `;

      const Comp3 = styled(Comp2)`
        text-align: center;
      `;

      expect(Comp.displayName).toMatchInlineSnapshot(`"styled.div"`);
      expect(Comp2.displayName).toMatchInlineSnapshot(`"Styled(styled.div)"`);
      expect(Comp3.displayName).toMatchInlineSnapshot(`"Styled(Styled(styled.div))"`);
      expect(render(<Comp />).container).toMatchInlineSnapshot(`
        <div>
          <div
            class="sc-a d"
          />
        </div>
      `);
      expect(render(<Comp2 />).container).toMatchInlineSnapshot(`
        <div>
          <div
            class="sc-a sc-b d e"
          />
        </div>
      `);
      expect(render(<Comp3 as="span" />).container).toMatchInlineSnapshot(`
        <div>
          <span
            class="sc-a sc-b sc-c d e f"
          />
        </div>
      `);

      expect(getRenderedCSS()).toMatchInlineSnapshot(`
        ".d {
          background: blue;
          color: red;
        }
        .e {
          color: green;
        }
        .f {
          text-align: center;
        }"
      `);
    });

    it('"as" prop signature should inform rendered JSX if provided', () => {
      const X = styled.div<{ as?: 'div' | 'button' }>``;
      const StyledX = styled(X)``;

      render(
        <>
          <X
            // @ts-expect-error invalid input test
            as="section"
          />
          <StyledX
            // @ts-expect-error invalid input test
            as="section"
          />
        </>
      );
    });
  });
});
