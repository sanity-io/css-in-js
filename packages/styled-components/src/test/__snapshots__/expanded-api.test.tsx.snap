// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`expanded api > "as" prop > changes the rendered element type 1`] = `
<div>
  <span
    class="sc-a b"
  />
</div>
`;

exports[`expanded api > "as" prop > changes the rendered element type when used with attrs 1`] = `
<div>
  <header
    class="sc-a b"
  />
</div>
`;

exports[`expanded api > "as" prop > works with custom components 1`] = `
<div>
  <figure
    class="sc-a b"
  />
</div>
`;

exports[`expanded api > chaining > should merge the options strings 1`] = `
<div>
  <div
    class="dn-2-id-1"
  />
</div>
`;

exports[`expanded api > componentId > should be attached if passed in 1`] = `
<div>
  <div
    class="LOLOMG"
  />
</div>
`;

exports[`expanded api > componentId > should be attached if passed in 2`] = `
<div>
  <div
    class="OMGLOL"
  />
</div>
`;

exports[`expanded api > componentId > should be combined with displayName if both passed in 1`] = `
<div>
  <div
    class="Comp-LOLOMG"
  />
</div>
`;

exports[`expanded api > componentId > should be combined with displayName if both passed in 2`] = `
<div>
  <div
    class="Comp2-OMGLOL"
  />
</div>
`;

exports[`expanded api > componentId > should be generated as "sc" + hash 1`] = `
<div>
  <div
    class="sc-a"
  />
</div>
`;

exports[`expanded api > componentId > should be generated as "sc" + hash 2`] = `
<div>
  <div
    class="sc-b"
  />
</div>
`;

exports[`expanded api > componentId > should be generated from displayName + hash 1`] = `
<div>
  <div
    class="Comp-a"
  />
</div>
`;

exports[`expanded api > componentId > should be generated from displayName + hash 2`] = `
<div>
  <div
    class="Comp2-b"
  />
</div>
`;
